bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/events/OSGiEventManager.java

```diff
@@ -91,21 +84,25 @@ public void removedService(ServiceReference reference, Object service) {
 
     }
 
-    private final Logger logger = LoggerFactory.getLogger(OSGiEventManager.class);
+    private final Map<String, EventFactory> typedEventFactories = new ConcurrentHashMap<String, EventFactory>();
 
-    private EventAdmin osgiEventAdmin;
+    private final ConcurrentHashMap<String, CopyOnWriteArraySet<EventSubscriber>> typedEventSubscribers = new ConcurrentHashMap<>();
 
-    private final Map<String, EventFactory> typedEventFactories = new ConcurrentHashMap<String, EventFactory>();
+    // private final SetMultimap<String, EventSubscriber> typedEventSubscribers = Multimaps
+    // .synchronizedSetMultimap(HashMultimap.<String, EventSubscriber> create());
 
-    private final SetMultimap<String, EventSubscriber> typedEventSubscribers = Multimaps
-            .synchronizedSetMultimap(HashMultimap.<String, EventSubscriber> create());
+    private ThreadedEventHandler eventHandler;
 
     private EventSubscriberServiceTracker eventSubscriberServiceTracker;
 
+    private EventAdmin osgiEventAdmin;
+
     private SafeCaller safeCaller;
 
     @Activate
     protected void activate(ComponentContext componentContext) {
+        eventHandler = new ThreadedEventHandler(typedEventSubscribers, typedEventFactories, safeCaller);
+
         eventSubscriberServiceTracker = new EventSubscriberServiceTracker(componentContext.getBundleContext());
```

Comment:
Slightly unrelated, but as you are refactoring the code anyhow in this PR: Do you happen to know, why we work with a service tracker here instead of using a 0..n dynamic dependency? I especially wonder in the light of #4775 whether we are doing anything wrong here...

Reply:
I need to guess, but AFAIK ESH used service trackers a lot "in the early days".
And IIRC the core should not use DS "in the early days".

We could migrate that from ServiceTracker to DS, but this would be another change and is not part of that "topic".

Reply:
ok.